pkg/querier/batch/merge.go

```diff
@@ -28,18 +28,30 @@ type mergeIterator struct {
 	currErr error
 }
 
-func newMergeIterator(cs []GenericChunk) *mergeIterator {
-	css := partitionChunks(cs)
-	its := make([]*nonOverlappingIterator, 0, len(css))
-	for _, cs := range css {
-		its = append(its, newNonOverlappingIterator(cs))
+func newMergeIterator(it iterator, gc []GenericChunk) *mergeIterator {
+	c, ok := it.(*mergeIterator)
+	if ok {
+		c.nextBatchBuf[0] = chunk.Batch{}
+		c.currErr = nil
+	} else {
+		c = &mergeIterator{}
 	}
 
-	c := &mergeIterator{
-		its:        its,
-		h:          make(iteratorHeap, 0, len(its)),
-		batches:    make(batchStream, 0, len(its)),
-		batchesBuf: make(batchStream, len(its)),
+	css := partitionChunks(gc)
+	putGenericChunkSlice(gc)
+	if cap(c.its) >= len(css) {
+		c.its = c.its[:len(css)]
+		c.h = c.h[:0]
+		c.batches = c.batches[:0]
+		c.batchesBuf = c.batchesBuf[:len(css)]
```

Comment:
[nit] I would add a comment explaining that we're not resetting the content of `c.batchesBuf` because they will be reset once reused later on by `mergeStreams()`.